### PR TITLE
Feature: Deposit Service

### PR DESCRIPTION
**Pull request by re1ro**
* [x] Refactoring: abstract `Proxy` and `Upgradable` for `GasService` and `DepositService`
* [x] `DepositService`: generating address and sending or unwrapping native currency and sending ERC20 tokens as well
* [x] Unit tests for `DepositService`

## Design overview
There are 3 main functions to this service: `sendToken`, `sendNative` and `withdrawNative`. Each of the functions are expecting the funds to be already transferred to a specific deposit address which is generated with following functions:
`depositAddressForSendToken`, `depositAddressForSendNative`, `depositAddressForWithdrawNative`. 

Here is a more detailed overview of the parameters they take:
```
    function depositAddressForSendToken(
        bytes32 nonce,
        string calldata destinationChain,
        string calldata destinationAddress,
        string calldata tokenSymbol
    ) external view returns (address);

    function sendToken(
        bytes32 nonce,
        string calldata destinationChain,
        string calldata destinationAddress,
        string calldata tokenSymbol
    ) external;
```

```
    function depositAddressForSendNative(
        bytes32 nonce,
        string calldata destinationChain,
        string calldata destinationAddress
    ) external view returns (address);

    function sendNative(
        bytes32 nonce,
        string calldata destinationChain,
        string calldata destinationAddress
    ) external;
```

```
    function depositAddressForWithdrawNative(bytes32 nonce, address recipient) 
        external view returns (address);

    function withdrawNative(bytes32 nonce, address payable recipient) external;
```

### CODE REVIEW

File: src/util/Proxy.sol [hunk 1]
```diff
@@ -2,20 +2,27 @@
...
 
     // bytes32(uint256(keccak256('eip1967.proxy.implementation')) - 1)
     bytes32 internal constant _IMPLEMENTATION_SLOT = 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
+    // keccak256('owner');
+    bytes32 internal constant _OWNER_SLOT = 0x02016836a56b71f0d02689e69e326f4f4c1b9057164ef592671cf0d37c8040c0;
 
-    constructor(address gasReceiverImplementation, bytes memory params) {
+    constructor(
+        address implementationAddress,
+        address owner,
```
[the remark this answers is not in the document]
Comment (re1ro) [reply src/util/Proxy.sol:16]: Will do

File: src/deposit-service/AxelarDepositService.sol [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        bytes32 salt = keccak256(abi.encode(PREFIX_DEPOSIT_TOKEN_UNWRAP, recipient, senderAddress));
+        return _depositAddress(salt);
+    }
+
+    function handleTokenSend(
+        string calldata destinationChain,
+        string calldata destinationAddress,
+        string calldata senderAddress,
+        string calldata tokenSymbol,
+        address tokenAddress
```
[the remark this answers is not in the document]
Comment (re1ro) [reply src/deposit-service/AxelarDepositService.sol:62]: Will change that

File: src/deposit-service/AxelarDepositService.sol [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        returns (address)
+    {
+        bytes32 salt = keccak256(abi.encode(PREFIX_DEPOSIT_TOKEN_UNWRAP, recipient, senderAddress));
+        return _depositAddress(salt);
+    }
+
+    function handleTokenSend(
+        string calldata destinationChain,
+        string calldata destinationAddress,
+        string calldata senderAddress,
```
[the remark this answers is not in the document]
Comment (re1ro) [reply src/deposit-service/AxelarDepositService.sol:60]: Should we use nonce `byted32`

File: src/deposit-service/AxelarDepositService.sol [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        DepositReceiver depositReceiver = new DepositReceiver{
+            salt: keccak256(
+                abi.encode(PREFIX_DEPOSIT_TOKEN_SEND, destinationChain, destinationAddress, senderAddress, tokenSymbol)
+            )
+        }();
+        uint256 amount = IERC20(tokenAddress).balanceOf(address(depositReceiver));
+
+        if (amount == 0) revert NothingDeposited();
+
+        (bool success, bytes memory returnData) = depositReceiver.execute(
```
[the remark this answers is not in the document]
Comment (re1ro) [reply src/deposit-service/AxelarDepositService.sol:73]: Yeah had issues with that but we made it work

File: src/deposit-service/AxelarDepositService.sol [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        address gatewayAddress = gateway();
+
+        IERC20(tokenAddress).approve(gatewayAddress, amount);
+        IAxelarGateway(gatewayAddress).sendToken(destinationChain, destinationAddress, tokenSymbol, amount);
+    }
+
+    function handleNativeSend(
+        string calldata destinationChain,
+        string calldata destinationAddress,
+        string calldata senderAddress
```
[the remark this answers is not in the document]
Comment (re1ro) [reply src/deposit-service/AxelarDepositService.sol:92]: Will change to nonce

File: src/deposit-service/AxelarDepositService.sol [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        address gatewayAddress = gateway();
+        address token = wrappedToken();
+        string memory symbol = wrappedSymbol();
+
+        IWETH9(token).deposit{ value: amount }();
+        IERC20(token).approve(gatewayAddress, amount);
+        IAxelarGateway(gatewayAddress).sendToken(destinationChain, destinationAddress, symbol, amount);
+    }
+
+    function handleTokenUnwrap(address payable recipient, string calldata senderAddress) external {
```
[the remark this answers is not in the document]
Comment (re1ro) [reply src/deposit-service/AxelarDepositService.sol:116]: For withdrawing native coin

File: src/deposit-service/AxelarDepositService.sol [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    function _setup(bytes calldata data) internal override {
+        (address gatewayAddress, string memory symbol) = abi.decode(data, (address, string));
+
+        if (gatewayAddress == address(0)) revert InvalidAddress();
+
+        if (IAxelarGateway(gatewayAddress).tokenAddresses(symbol) == address(0)) revert InvalidSymbol();
+
+        bytes memory symbolBytes = bytes(symbol);
+
+        if (symbolBytes.length == 0 || symbolBytes.length > 30) revert InvalidSymbol();
```
Comment (sammy1991106) [inline src/deposit-service/AxelarDepositService.sol:228]: @sergeynog Do you see we ever use symbol with length >30 bytes?

**Review by sammy1991106 — COMMENTED** [review]
Please also go through DepositService again to extract repeated code into some functions.

File: src/deposit-service/AxelarDepositService.sol [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+        if (amount == 0) revert NothingDeposited();
+
+        (bool success, bytes memory returnData) = depositReceiver.execute(
+            tokenAddress,
+            0,
+            abi.encodeWithSelector(IERC20.approve.selector, gatewayAddress, amount)
+        );
+
+        if (!success || (returnData.length != uint256(0) && !abi.decode(returnData, (bool)))) revert TransferFailed();
```
Comment (sammy1991106) [inline src/deposit-service/AxelarDepositService.sol:75]:
```suggestion
        if (!success || (returnData.length != uint256(0) && !abi.decode(returnData, (bool)))) revert ApproveFailed();
```

File: src/deposit-service/DepositReceiver.sol [hunk 9]
```diff
@@ -2,8 +2,34 @@
...
+        if (_owner != msg.sender) revert NotOwner();
+        _;
+    }
+
+    function execute(
+        address callee,
+        uint256 value,
+        bytes calldata data
+    ) external onlyOwner returns (bool success, bytes memory returnData) {
+        if (callee.code.length != 0) {
```
[the remark this answers is not in the document]
Comment (re1ro) [reply src/deposit-service/DepositReceiver.sol:24]: We should revert here, agreed

[the remark this answers is not in the document]
Comment (re1ro) [reply src/deposit-service/DepositReceiver.sol:24]: And old deposit handler as well

File: src/deposit-service/DepositReceiver.sol [hunk 10]
```diff
@@ -2,8 +2,34 @@
...
+    modifier onlyOwner() {
+        if (_owner != msg.sender) revert NotOwner();
+        _;
+    }
+
+    function execute(
+        address callee,
+        uint256 value,
+        bytes calldata data
+    ) external onlyOwner returns (bool success, bytes memory returnData) {
```
[the remark this answers is not in the document]
Comment (re1ro) [reply src/deposit-service/DepositReceiver.sol:23]: Try to write a unitest for reentrancy protection

File: src/deposit-service/AxelarDepositService.sol (hunk 8 above)
Comment (re1ro) [reply src/deposit-service/AxelarDepositService.sol:75]: Will add more errors

**Review by sammy1991106 — APPROVED** [review]

**Pull request merged**